File: wear/src/main/java/com/asamm/locus/addon/wear/gui/custom/WaypointInputTextActivity.java
/*
 * Created by milan on 01.08.2018.
 * This code is part of Locus project from Asamm Software, s. r. o.
 * Copyright (C) 2018
 */
package com.asamm.locus.addon.wear.gui.custom;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.speech.RecognizerIntent;
import android.support.annotation.Nullable;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;

import com.asamm.locus.addon.wear.R;
import com.asamm.locus.addon.wear.common.communication.DataPath;
import com.asamm.locus.addon.wear.common.communication.containers.DataPayload;
import com.asamm.locus.addon.wear.gui.LocusWearActivity;
import com.asamm.locus.addon.wear.gui.LocusWearActivityHwKeyDelegate;
import com.asamm.locus.addon.wear.utils.UtilsGui;

import java.util.List;

import locus.api.utils.Logger;

/**
 * Text input activity to input new waypoint name
 */
public class WaypointInputTextActivity extends LocusWearActivity {
    public static final String KEY_RESULT_DATA = "KEY_RESULT_DATA";
    private static final int REQUEST_SPEECH = 2;

    @Override
    protected DataPayload getInitialCommandType() {
        return null;
    }

    @Override
    protected DataPath getInitialCommandResponseType() {
        return null;
    }

    @Override
    public void registerHwKeyActions(LocusWearActivityHwKeyDelegate delegate) {
    }

    @Override
    protected boolean isMakeHandshakeOnStart() {
        return false;
    }

    @Override
    public boolean isChildLocusWearActivity() {
        return true;
    }

    private EditText editText;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_waypoint_text_input);

        editText = findViewById(R.id.edit_text);
        editText.setOnEditorActionListener((v, actionId, event) -> {
            UtilsGui.hideKeyboard(editText);
            finishWithResult(v.getText().toString());
            return false;
        });

        Button btnKeyboard = findViewById(R.id.btn_keyboard);
        btnKeyboard.setOnClickListener(v -> {
            // show dummy 1x1px sized EditText and open keyboard
            editText.setVisibility(View.VISIBLE);
            UtilsGui.showKeyboard(editText);
        });
        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        int numInputMethods = imm.getEnabledInputMethodList().size();
        Logger.logD("KEYBOARD", "num input methods: "+numInputMethods);
        if ( numInputMethods == 0) {
            btnKeyboard.setVisibility(View.GONE);
            editText.setVisibility(View.GONE);
        }

        Button btnSpeech = findViewById(R.id.btn_mic);
        btnSpeech.setOnClickListener(v -> displaySpeechRecognizer());

        Button btnDefault = findViewById(R.id.btn_ok);
        btnDefault.setOnClickListener(v -> finishWithResult(""));
    }

    // Create an intent that can start the Speech Recognizer activity
    private void displaySpeechRecognizer() {
        Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL,
                RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
        // Start the activity, the intent will be populated with the speech text
        startActivityForResult(intent, REQUEST_SPEECH);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_SPEECH && resultCode == RESULT_OK) {
            List<String> results = data.getStringArrayListExtra(
                    RecognizerIntent.EXTRA_RESULTS);
            String spokenText = results.get(0);
            finishWithResult(spokenText);
        }
    }

    private void finishWithResult(String result) {
        Intent resultIntent = new Intent();
        resultIntent.putExtra(KEY_RESULT_DATA, result);
        setResult(RESULT_OK, resultIntent);
        finish();
    }
}

File: wear/src/main/java/com/asamm/locus/addon/wear/gui/custom/WaypointInputTextActivity.kt
/*
 * Created by milan on 01.08.2018.
 * This code is part of Locus project from Asamm Software, s. r. o.
 * Copyright (C) 2018
 */
package com.asamm.locus.addon.wear.gui.custom

import android.content.Intent
import android.os.Bundle
import android.speech.RecognizerIntent
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import com.asamm.locus.addon.wear.R
import com.asamm.locus.addon.wear.common.communication.DataPath
import com.asamm.locus.addon.wear.common.communication.containers.DataPayload
import com.asamm.locus.addon.wear.gui.LocusWearActivity
import com.asamm.locus.addon.wear.gui.UiUtils.hideKeyboard
import com.asamm.locus.addon.wear.gui.UiUtils.showKeyboard

/**
 * Text input activity to input new waypoint name
 */
class WaypointInputTextActivity : LocusWearActivity() {

    override val initialCommandType: DataPayload<*>?
        get() = null

    override val initialCommandResponseType: DataPath?
        get() = null

    override val isMakeHandshakeOnStart: Boolean
        get() = false

    override val isChildLocusWearActivity: Boolean
        get() = true

    private lateinit var editText: EditText

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_waypoint_text_input)
        editText = findViewById(R.id.edit_text)
        editText.setOnEditorActionListener { v: TextView, _: Int, _: KeyEvent? ->
            hideKeyboard(editText)
            finishWithResult(v.text.toString())
            true
        }
        val btnKeyboard = findViewById<Button>(R.id.btn_keyboard)
        btnKeyboard.setOnClickListener {
            showKeyboard(editText)
        }
        val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
        val numInputMethods = imm.enabledInputMethodList.size
        if (numInputMethods == 0) {
            btnKeyboard.visibility = View.GONE
        }
        val btnSpeech = findViewById<Button>(R.id.btn_mic)
        btnSpeech.setOnClickListener { displaySpeechRecognizer() }
        val btnDefault = findViewById<Button>(R.id.btn_ok)
        btnDefault.setOnClickListener { finishWithResult("") }
    }

    // Create an intent that can start the Speech Recognizer activity
    private fun displaySpeechRecognizer() {
        val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)
        intent.putExtra(
                RecognizerIntent.EXTRA_LANGUAGE_MODEL,
                RecognizerIntent.LANGUAGE_MODEL_FREE_FORM
        )
        // Start the activity, the intent will be populated with the speech text
        startActivityForResult(intent, REQUEST_SPEECH)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_SPEECH && resultCode == RESULT_OK) {
            val results: List<String>? = data!!.getStringArrayListExtra(
                    RecognizerIntent.EXTRA_RESULTS
            )
            val spokenText = results!![0]
            finishWithResult(spokenText)
        }
    }

    private fun finishWithResult(result: String) {
        val resultIntent = Intent()
        resultIntent.putExtra(KEY_RESULT_DATA, result)
        setResult(RESULT_OK, resultIntent)
        finish()
    }

    companion object {

        const val KEY_RESULT_DATA = "KEY_RESULT_DATA"

        private const val REQUEST_SPEECH = 2
    }
}